src/Nerdbank.MessagePack/Converters/HardwareAccelerated.cs

```diff
@@ -0,0 +1,2634 @@
+		/// All subsequence elements must be contiguous in memory after the first.
+		/// This buffer must be at least <paramref name="inputLength"/> elements long.
+		/// </param>
+		/// <param name="inputLength">The number of elements to encode.</param>
+		/// <returns>The number of msgpack bytes written.</returns>
+		private static nuint Write(ref byte output, ref bool values, nuint inputLength)
+		{
+			nuint offset = 0;
+			if (Vector.IsHardwareAccelerated && inputLength >= unchecked((nuint)Vector<sbyte>.Count))
+			{
+				for (; offset + unchecked((nuint)Vector<sbyte>.Count) <= inputLength; offset += unchecked((nuint)Vector<sbyte>.Count))
+				{
+					Vector<sbyte> results = Vector.Equals(Vector.LoadUnsafe(ref Unsafe.As<bool, sbyte>(ref values), offset), Vector<sbyte>.Zero) + new Vector<sbyte>(unchecked((sbyte)MessagePackCode.True));
+					results.StoreUnsafe(ref Unsafe.As<byte, sbyte>(ref output), offset);
+				}
+
+				{
+					offset = inputLength - unchecked((nuint)Vector<sbyte>.Count);
+					Vector<sbyte> results = Vector.Equals(Vector.LoadUnsafe(ref Unsafe.As<bool, sbyte>(ref values), offset), Vector<sbyte>.Zero) + new Vector<sbyte>(unchecked((sbyte)MessagePackCode.True));
+					results.StoreUnsafe(ref Unsafe.As<byte, sbyte>(ref output), offset);
+				}
+			}
+			else
+			{
+				for (; offset < inputLength; offset++)
+				{
+					Unsafe.Add(ref output, offset) = Unsafe.Add(ref values, offset) ? MessagePackCode.True : MessagePackCode.False;
+				}
+			}
+
+			return inputLength;
+		}
+
+		/// <inheritdoc cref="Write(ref byte, ref bool, nuint)"/>
+		private static nuint Write(ref byte output, ref sbyte input, nuint inputLength)
+		{
+			nuint outputOffset = 0;
+			for (nuint inputOffset = 0; inputOffset < inputLength; inputOffset++)
+			{
+				sbyte temp = Unsafe.Add(ref input, inputOffset);
+				if (temp < MessagePackRange.MinFixNegativeInt)
+				{
+					Unsafe.Add(ref output, outputOffset++) = MessagePackCode.Int8;
+				}
+
+				Unsafe.Add(ref output, outputOffset++) = unchecked((byte)temp);
+			}
+
+			return outputOffset;
+		}
+
+		private static DecodeResult Read(ref ushort output, int outputCapacity, in byte msgpack, int inputLength, out int writtenLength, out int readLength)
+		{
+			ref byte input = ref Unsafe.AsRef(in msgpack);
+			writtenLength = 0;
+			readLength = 0;
+
+			for (; readLength < inputLength && writtenLength < outputCapacity;)
+			{
+				var code = Unsafe.Add(ref input, readLength++);
+				if (unchecked((sbyte)code) >= MessagePackRange.MinFixNegativeInt)
+				{
+					Unsafe.Add(ref output, writtenLength++) = code;
+					continue;
+				}
+
+				switch (code)
+				{
+					case MessagePackCode.UInt8:
+						if (readLength >= inputLength)
+						{
+							--readLength;
+							return DecodeResult.InsufficientBuffer;
+						}
+
+						Unsafe.Add(ref output, writtenLength++) = Interpret<byte, ushort>(ref input, readLength);
+						readLength += sizeof(byte);
+						break;
+					case MessagePackCode.UInt16:
```

Comment:
There is a very large amount of code in this file. When I ran the benchmarks on my Ryzen 9 machine, deserialization was not improved for the integer types except for `Int8`. If you see the same, I wonder if we can drop a lot of the code that is redundant with our more conventional read converters for integers except for Int8. I'm feeling like "less is more" when it comes to code because it's easier to test and maintain. So we should keep it only if it's substantially better (or if we have reason to believe future chips will make it better and we want to keep the work you've done in the meantime).
BTW with regard to `Int8[]` (aka `sbyte[]`) I'm thinking we should just encode that using msgpack binary, the same as we do for the unsigned `byte[]`. I see very little lost, and what's lost is only if the deserializer doesn't even know the type they are expecting to deserialize into. But they'll gain an enormous perf boost in the process.

Thoughts?

Reply:
Treating `sbyte[]` like `byte[]` will introduce a lot of performance improvement.
Interoperability between languages will decrease. If there is going to be a problem, it will be with ASP.NET.
I do not have objection due to the benefit.

Reply:
> Interoperability between languages will decrease. If there is going to be a problem, it will be with ASP.NET.

Can you elaborate on that? I want to make sure I'm not missing something. I'm thinking that deserializing a payload from another language should be just fine because C# can just deserialize as `byte[]` if that's what the other language intended.
I guess _serializing_ an `sbyte[]` may lead to other libraries/platforms misinterpreting it as `byte[]` leading to the signed integers being misinterpreted as unsigned. I guess if that's a concern we could add an option to switch back to serializing as an array of sbyte msgpack values.
What would really help is to know what the use case is for serializing an `sbyte[] `in the first place is, because I personally never used `sbyte`. Is this interoperability trade off for perf likely to pay off more than it hurts?

Reply:
I agree that using `sbyte[]` is very rare.

```csharp
enum Relationship : sbyte
{
  Banned = -1, // Comparing with 0 is efficient than with not 0 number.
  None = 0,
  Friendship = 1,
}
```

The above uses `enum` backed by `sbyte` (I sometimes write this kind of codes). If one uses the programming style of [Structure of Array](https://en.wikipedia.org/wiki/AOS_and_SOA), then they uses `sbyte[]`.
The game programmers sometimes adopt the above `SoA` programming style (See: [Unity Entity Component System](https://docs.unity3d.com/Packages/com.unity.entities@1.0/manual/index.html)).
Yeah this example is very very rare and will not occur for most programmers.
Those who use `sbyte[]` in cross-language env know what they are doing. It is enough to give them an option to switch back to `sbyte array` and warnings.